libcodechecker/server/database/database.py

```diff
@@ -53,62 +58,224 @@ def __init__(self, model_meta, migration_root):
+            database_schema_revision = mcontext.get_current_revision()
+            LOG.debug("Schema revision in the database: " +
+                      str(database_schema_revision))
+
+            if database_schema_revision is None:
+                LOG.debug("Database schema should have been created!")
+                return DBStatus.SCHEMA_MISSING
+
+            LOG.debug("Checking schema versions in the package.")
+            schema_config_head = scriptt.get_current_head()
+
+            if database_schema_revision != schema_config_head:
+                LOG.debug("Database schema mismatch detected "
+                          "between the package and the database")
+                LOG.debug("Checking if automatic upgrade is possible.")
+                all_revs = [rev.revision for rev in
+                            scriptt.walk_revisions()]
+
+                if database_schema_revision not in all_revs:
+                    LOG.debug("Automatic schema upgrade is not possible!")
+                    LOG.debug("Please re-check your database and"
+                              "CodeChecker versions!")
+                    return DBStatus.SCHEMA_MISMATCH_NO
+
+                # There is a schema mismatch.
+                return DBStatus.SCHEMA_MISMATCH_OK
+            else:
+                LOG.debug("Schema in the package and"
+                          " in the database is the same.")
+                LOG.debug("No schema modification is needed.")
+                return DBStatus.OK
+
+        except sqlalchemy.exc.SQLAlchemyError as alch_err:
+            LOG.debug(str(alch_err))
+            return DBStatus.FAILED_TO_CONNECT
+        except CommandError as cerr:
+            LOG.debug("Database schema and CodeChecker is incompatible. "
+                      "Please update CodeChecker.")
+            LOG.debug(str(cerr))
+            return DBStatus.SCHEMA_MISMATCH
+
+        finally:
+            session.close()
+            connection.close()
+            engine.dispose()
+
+    def upgrade(self):
+        """
+        Upgrade database db schema.
 
-        If use_migration is True, this method runs an alembic upgrade to HEAD.
+        Checks the database schema for schema mismatch.
+        The database server has to be started before this method is called.
 
-        In the False case, there is no migration support and only SQLAlchemy
-        meta data is used for schema creation.
+        This method runs an alembic upgrade to HEAD.
 
-        On error sys.exit(1) is called.
         """
 
+        # another safety check before we initialize or upgrade the schema
+        ret = self.check_schema()
+
+        migration_ok = [DBStatus.SCHEMA_MISMATCH_OK,
+                        DBStatus.SCHEMA_MISSING]
+        if ret not in migration_ok:
+            # schema migration is not possible
+            return ret
+
+        try:
+            engine = self.create_engine()
```

Comment:
This pattern of getting the engine, session, connection is very frequent. Maybe it is worth to have a utility function for this?